jax/experimental/host_callback.py

```diff
@@ -148,25 +220,89 @@ def power3(x):
 See documentation for :func:`id_tap` and :func:`id_print`.
 For more usage example, see tests/host_callback_test.py.
 
-Still to do:
-  * Performance tests.
-  * Add flags for logging.
-  * Add unit tests with mocks.
-  * Explore implementation with outside compilation.
-  * Explore an extended API that allows the host function to return
-    values to the accelerator computation.
-
-
 Low-level details and debugging
 -------------------------------
 
+The host callback functions will be executed for each device in the order in which
+the send operations were performed on the device.
+
+The host callback functions for multiple devices may be interleaved.
+The data from the devices is received by separate threads managed by the JAX
+runtime (one thread per device). The runtime maintains a buffer of
+configurable size. When the buffer is full, all the receiving threads are paused
+which eventually pauses the computation on devices. The runtime has one
+additional thread that invokes the Python user functions with the received data.
+If the processing of the callbacks is slow, it may actually lead to the runtime
+buffer filling up, and eventually pausing the computation on the devices
+when they need to send something. For more details on the outfeed receiver
+runtime mechanism see
+`runtime code
+<https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/python/outfeed_receiver.cc>`_.
+
+In order to pause the execution until all data from computations already
+started on devices has arrived and has been processed, use :func:`barrier_wait`.
+Note that this is needed only for :func:`id_tap` and :func:`id_print`, which
+are processed asyncronously with the device computation.
+
+Exceptions from the user-defined callback functions are logged along with their
+stack traces, but the receiving threads are not stopped. Instead the last
+exception is recorded and the subsequent :func:`barrier_wait` will
+raise :exc:`TapFunctionException` if any exception had occurred
+in one of the tap functions. This exception will include the text and the
+stack trace of the last exception encountered.
+
+One further complication arises for callback functions that must return
```

Comment:
I'd be curious if anybody has a better idea for how to handle errors in the callback functions

Reply:
It seems like this is something that should be explicitly supported in XLA.